Add support for Windows

jillguyonnet opened the pull request:
## Issue description

EPR currently fails to run on Windows due to the use of `filepath` library in an abstracted file system.

## What does this PR do?

* Add a Jenkins testing stage on Windows
* Fix Windows compatiblity

## Why is it important?

EPR currently cannot run on Windows.

## Related issues

* Closes https://github.com/elastic/package-registry/issues/955

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-956/detail/PR-956/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-956/detail/PR-956/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-956/detail/PR-956/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-registry%2FPR-956/detail/PR-956/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://package-registry_956.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-02-08T08:50:48.205Z&rangeTo=2023-02-08T09:10:48.205Z&transactionName=BUILD+Ingest-manager%2Fpackage-registry%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=1c40ebdbc280bb66a8c341558fdb1b4e&transactionId=fb02ba1f73363939)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-02-08T09:00:48.205+0000


* Duration: 5 min 4 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 244  |
| Skipped      | 0 |
| Total        | 244   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

mrodm commented on .ci/Jenkinsfile:
```diff
@@ -71,11 +71,28 @@ pipeline {
+            cleanup()
+            withMageEnv(){
+              dir("${BASE_DIR}"){
+                sh(label: 'Runs the (unit) tests',script: 'mage -debug test|tee tests-report.txt')
```
As this is run in Windows, it should be called `bat`
it also seems it can be called `cmd` so it uses the right call under the hood.
https://github.com/elastic/apm-pipeline-library/blob/main/vars/cmd.groovy

jsoriano commented on packages/fs.go:
```diff
@@ -35,23 +36,29 @@ func NewExtractedPackageFileSystem(p *Package) (*ExtractedPackageFileSystem, err
 
 func (fs *ExtractedPackageFileSystem) Stat(name string) (os.FileInfo, error) {
-	return os.Stat(filepath.Join(fs.path, name))
+	return os.Stat(path.Join(fs.path, name))
```
I think that when accessing the file system directly, we should keep using `filepath`, that takes into account the different separator in Windows. This would be for all methods of `ExtractedPackageFileSystem`, that work on the native file system.

As a rule of thumb, if the path is used with a function of `os` package, it should be manipulated with `filepath` package.

jsoriano replied:
If we want to work with `path`, we could use [`os.DirFS`](https://pkg.go.dev/os#DirFS) to build a `fs.FS`.

jsoriano commented on packages/fs.go:
```diff
@@ -35,23 +36,29 @@ func NewExtractedPackageFileSystem(p *Package) (*ExtractedPackageFileSystem, err
 		return
 	}
 	for i := range matches {
-		match, err := filepath.Rel(fs.path, matches[i])
```
If we keep using `filepath` here, we would be able to continue using `filepath.Rel`.

jsoriano commented on packages/fs.go:
```diff
@@ -73,8 +80,8 @@ func NewZipPackageFileSystem(p *Package) (*ZipPackageFileSystem, error) {
-		name := filepath.Clean(f.Name)
-		parts := strings.Split(name, string(filepath.Separator))
+		name := path.Clean(f.Name)
+		parts := strings.Split(name, "/")
 		if len(parts) == 2 && parts[1] == "manifest.yml" {
```
Nit. Maybe we can use path functions here instead of manipulating the path as a string.
```suggestion
		dir, base := path.Split(name)
		if path.Dir(dir) == "." && base == "manifest.yml" {
```

jillguyonnet replied:
@jsoriano Seems the suggested nit doesn't work as is. I'm not exactly sure what the original intent is, but in the tests the condition
```go
if len(parts) == 2 && parts[1] == "manifest.yml"
```
is only true when the given path is `"example-1.0.1/manifest.yml"`. In this case, `path.Dir(dir)` is equal to `"example-1.0.1"`.

The other case where the file name is equal to `"manifest.yml"` is when the given path is `"example-1.0.1/data_stream/foo/manifest.yml"`, in which case `path.Dir(dir)` is equal to `"example-1.0.1/data_stream/foo"`.

There are no cases where `path.Dir(dir)` is equal to `"."` and the file name is equal to `"manifest.yml"`.

Please let me know if I'm missing something.

jillguyonnet replied:
Thank you for the explanations and playground! I would be inclined to add comment to make the intent explicit, thereby clarifying the partial condition `dir != "." && path.Dir(dir) == "."`. Knowing this, I prefer this option.

jillguyonnet commented:
Hi @jsoriano thanks for the tips, I'm on a learning curve so appreciate it!

jsoriano commented:
Hey @jillguyonnet,

I am thinking that it could be easier to fix the pending issues if we use `path` everywhere, as you were originally doing. This would probably solve a couple of issues:
* The paths in the API that are currently causing the tests to fail.
* Methods of the `ExtractedPackageFileSystem` type would always work with POSIX paths, also when running on Windows.

For that, we could create the `ExtractedPackageFileSystem`s this way:
```
type ExtractedPackageFileSystem struct {
	path string
	root fs.FS
}

func NewExtractedPackageFileSystem(p *Package) (*ExtractedPackageFileSystem, error) {
	return &ExtractedPackageFileSystem{path: p.BasePath, root: os.DirFS(p.BasePath)}, nil
}
```
And we would make all operations with `path` and `fs` methods over `root`.

jillguyonnet commented:
/test

jsoriano commented on main_test.go:
```diff
@@ -574,7 +574,7 @@ func listArchivedFiles(t *testing.T, body []byte) []byte {
 
 	for _, f := range zipReader.File {
-		listing.WriteString(fmt.Sprintf("%d %s\n", f.UncompressedSize64, f.Name))
+		listing.WriteString(fmt.Sprintf("%d %s\n", f.UncompressedSize64, filepath.ToSlash(f.Name)))
```
Please add a comment here to remember that we cannot trust the format of `f.Name` and then we need to use `ToSlash`.

jsoriano commented on packages/datastream.go:
```diff
@@ -177,16 +177,16 @@ func NewDataStream(basePath string, p *Package) (*DataStream, error) {
-	pipelineDir := filepath.Join(d.BasePath, "elasticsearch", DirIngestPipeline)
-	paths, err := fs.Glob(filepath.Join(pipelineDir, "*"))
+	pipelineDir := path.Join(d.BasePath, "elasticsearch", DirIngestPipeline)
+	paths, err := fs.Glob(path.Join(pipelineDir, "*"))
```
Nit. Maybe we should start using `fsys` in general as abbreviation for filesystems to distingish from `fs` package.

jsoriano commented on packages/fs.go:
```diff
@@ -28,32 +28,47 @@ type PackageFileSystem interface {
+}
+
+func (fsys *ExtractedPackageFileSystem) Glob(pattern string) (matches []string, err error) {
+	fs.WalkDir(fsys.root, ".", func(p string, d fs.DirEntry, err error) error {
```
If I recall correctly, `WalkDir` returns an error, we should handle it.

jsoriano commented on packages/fs.go:
```diff
@@ -28,32 +28,47 @@ type PackageFileSystem interface {
 // ExtractedPackageFileSystem provides utils to access files in an extracted package.
 type ExtractedPackageFileSystem struct {
 	path string
```
I think the only reason to keep the `path` now here would be to use it in error messages so users know to what package in the filesystem we are referring to. Use it when displaying paths in error messages of this package.

jsoriano commented on packages/fs.go:
```diff
@@ -28,32 +28,47 @@ type PackageFileSystem interface {
 	}
-	for i := range matches {
-		match, err := filepath.Rel(fs.path, matches[i])
+	pf, ok := f.(PackageFile)
```
Please add a comment here mentioning that we expect this to be a plain file, that implements this interface. It may look a bit random to convert to this interface otherwise.

jsoriano commented on packages/fs.go:
```diff
@@ -28,32 +28,47 @@ type PackageFileSystem interface {
-		match, err := filepath.Rel(fs.path, matches[i])
+	pf, ok := f.(PackageFile)
+	if !ok {
+		defer f.Close()
+		return nil, fmt.Errorf("file does not implement PackageFile interface: %q", name)
```
We could think to even `panic()` here, as all the files here should actually implement `PackageFile`. But let's stay on the safer side and keep it as an error.

jsoriano replied:
This is a place where we may want to use `fsys.path` to display the full path to the user. Something like this:
```suggestion
		defer f.Close()
		return nil, fmt.Errorf("file does not implement PackageFile interface: %q", filepath.Join(f.path, filepath.FromSlash(name)))
```

If we do this on multiple places, you may consider to implement a helper method for that, something like this:
```
func (fsys *ExtractedPackageFileSystem) fullPath(name string) (string) {
	return filepath.Join(f.path, filepath.FromSlash(name))
}
```

jsoriano reviewed (commented):
Good job here, there were several tricky cases! Added some nitpicking.

I think the only thing I would like to see reviewed before merging would be to handle the error of `WalkDir`, and to include the `fsys.path` in error messages. For the rest it LGTM.

jsoriano commented on packages/fs.go:
```diff
@@ -28,31 +29,52 @@ type PackageFileSystem interface {
+}
+
+func (fsys *ExtractedPackageFileSystem) Glob(pattern string) (matches []string, err error) {
+	e := fs.WalkDir(fsys.root, ".", func(p string, d fs.DirEntry, err error) error {
```
Nit. Error variables in go are usually called `err`.
```suggestion
	err := fs.WalkDir(fsys.root, ".", func(p string, d fs.DirEntry, err error) error {
```
Or you can probably reuse the one already defined in the function.
```suggestion
	err = fs.WalkDir(fsys.root, ".", func(p string, d fs.DirEntry, err error) error {
```
In cases where you need to distinguish between different errors in the same scope, you can use `Err` as suffix.
```suggestion
	walkErr := fs.WalkDir(fsys.root, ".", func(p string, d fs.DirEntry, err error) error {
```

jillguyonnet commented:
@jsoriano Thanks for the review, I addressed your comments. Very good point about `fsys.path`. I edited a bit the error messages inside
```go
func (fsys *ExtractedPackageFileSystem) Glob(pattern string) (matches []string, err error)
```
and the only place I can see that might make use of `fsys.path` would be the final error returned by `WalkDir`, so currently:
```go
fmt.Errorf("failed to obtain path under package root path (%s): %w", pattern, e)
```
I suppose there could also be a custom error message in the `Open` method.

Can you please let me know your thoughts on this and of anything else I might have missed? Thanks!

jsoriano commented on packages/fs.go:
```diff
@@ -73,10 +95,12 @@ func NewZipPackageFileSystem(p *Package) (*ZipPackageFileSystem, error) {
+		dir = path.Clean(dir)
+		// Find manifest files at the root of the package
+		if dir != "." && path.Dir(dir) == "." && base == "manifest.yml" {
+			root = dir
 			found = true
```
Nit. Not related to this change, but `found` is redundant here, right? We can also check if `root != ""`.

jillguyonnet replied:
True, I had assumed that `found` was used for extra readability.

mrodm reviewed (approved):
Great job here!
